opentelemetry-sdk/benches/log_exporter.rs

```diff
@@ -0,0 +1,178 @@
+/*
+    The benchmark results:
+    criterion = "0.5.1"
+    OS: Ubuntu 22.04.3 LTS (5.15.146.1-microsoft-standard-WSL2)
+    Hardware: AMD EPYC 7763 64-Core Processor - 2.44 GHz, 16vCPUs,
+    RAM: 64.0 GB
+    | Test                           | Average time|
+    |--------------------------------|-------------|
+    | LogExporterWithFuture          | 280 ns      |
+    | LogExporterWithoutFuture       | 255 ns      |
+*/
+
+use std::sync::Mutex;
+use std::time::SystemTime;
+
+use async_trait::async_trait;
+use criterion::{criterion_group, criterion_main, Criterion};
+
+use opentelemetry::logs::{LogRecord as _, LogResult, Logger as _, LoggerProvider as _, Severity};
+
+use opentelemetry_sdk::export::logs::LogData;
+use opentelemetry_sdk::logs::LogProcessor;
+use opentelemetry_sdk::logs::LoggerProvider;
+use pprof::criterion::{Output, PProfProfiler};
+use std::fmt::Debug;
+
+// Run this benchmark with:
+// cargo bench --bench log_exporter
+#[async_trait]
+pub trait LogExporterWithFuture: Send + Sync + Debug {
+    async fn export(&mut self, batch: Vec<LogData>);
+}
+
+pub trait LogExporterWithoutFuture: Send + Sync + Debug {
+    fn export(&mut self, batch: Vec<LogData>);
+}
+
+#[derive(Debug)]
+struct NoOpExporterWithFuture {}
+
+#[async_trait]
+impl LogExporterWithFuture for NoOpExporterWithFuture {
+    async fn export(&mut self, _batch: Vec<LogData>) {}
+}
+
+#[derive(Debug)]
+struct NoOpExporterWithoutFuture {}
+impl LogExporterWithoutFuture for NoOpExporterWithoutFuture {
+    fn export(&mut self, _batch: Vec<LogData>) {}
+}
+
+#[derive(Debug)]
+struct ExportingProcessorWithFuture {
+    exporter: Mutex<NoOpExporterWithFuture>,
+}
+
+impl ExportingProcessorWithFuture {
+    fn new(exporter: NoOpExporterWithFuture) -> Self {
+        ExportingProcessorWithFuture {
+            exporter: Mutex::new(exporter),
+        }
+    }
+}
+
+impl LogProcessor for ExportingProcessorWithFuture {
+    fn emit(&self, data: &mut LogData) {
+        let mut exporter = self.exporter.lock().expect("lock error");
+        futures_executor::block_on(exporter.export(vec![data.clone()]));
```

Comment:
Do we block on runtime our LogProcrssor? It's generally not really performat

Reply:
We could try https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.block_on and see if there is difference. Maybe spin up the runtime as part of the Processor so we can reuse them

Reply:
+1. This is not testing the async runtime. Better to create the async runtime outside of the timing loop, and then block using tokio runtime.

Reply:
Modified to use Tokio runtime's `block_on`, and  perf drastically dropped. (280ns to 350 ns)
```
#[derive(Debug)]
struct ExportingProcessorWithFuture {
    exporter: Mutex<NoOpExporterWithFuture>,
    rt: Runtime,
}

impl ExportingProcessorWithFuture {
    fn new(exporter: NoOpExporterWithFuture) -> Self {
        let rt = Runtime::new().unwrap();
        ExportingProcessorWithFuture {
            exporter: Mutex::new(exporter),
            rt,      
        }
    }
}

impl LogProcessor for ExportingProcessorWithFuture {
    fn emit(&self, data: &mut LogData) {
        let mut exporter = self.exporter.lock().expect("lock error");
        // futures_executor::block_on(exporter.export(vec![data.clone()]));
        self.rt.block_on(exporter.export(vec![data.clone()]));
    }

    fn force_flush(&self) -> LogResult<()> {
        Ok(())
    }

    fn shutdown(&self) -> LogResult<()> {
        Ok(())
    }
}
```